otto/internal/database/repository.go

```diff
@@ -0,0 +1,327 @@
+	Delete(ctx context.Context, id string) error
+}
+
+// SQLMapper provides functions to map between entities and database rows.
+type SQLMapper[T Entity] interface {
+	// ToRow maps an entity to database column values.
+	ToRow(entity *T) []interface{}
+	// FromRow creates an entity from a database row.
+	FromRow(row *sql.Row) (*T, error)
+	// FromRows creates entities from database rows.
+	FromRows(rows *sql.Rows) ([]T, error)
+	// Columns returns the database column names.
+	Columns() []string
+	// Table returns the database table name.
+	Table() string
+}
+
+// SQLRepository implements Repository using SQL operations.
+type SQLRepository[T Entity] struct {
+	db     Provider
+	mapper SQLMapper[T]
+}
+
+// Ensure SQLRepository implements Repository.
+var _ Repository[Entity] = (*SQLRepository[Entity])(nil)
+
+// validateTableName ensures table names are safe for use in SQL queries.
+func validateTableName(tableName string) error {
+	// Restrict table names to alphanumeric characters, underscores
+	validPattern := regexp.MustCompile(`^[a-zA-Z0-9_]+$`)
+	if !validPattern.MatchString(tableName) {
+		return fmt.Errorf("invalid table name: %s", tableName)
+	}
+	return nil
+}
+
+// NewSQLRepository creates a new SQL repository.
+func NewSQLRepository[T Entity](db Provider, mapper SQLMapper[T]) (Repository[T], error) {
+	// Validate table name at initialization time
+	tableName := mapper.Table()
+	if err := validateTableName(tableName); err != nil {
+		return nil, fmt.Errorf("failed to create repository: %w", err)
+	}
+
+	return &SQLRepository[T]{
+		db:     db,
+		mapper: mapper,
+	}, nil
+}
+
+// FindByID retrieves an entity by its ID.
+func (r *SQLRepository[T]) FindByID(ctx context.Context, id string) (*T, error) {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	columns := r.mapper.Columns()
+	// #nosec G202 -- Table name is validated at initialization
+	query := "SELECT " + joinColumns(columns) + " FROM " + r.mapper.Table() + " WHERE id = ?"
+
+	row := r.db.DB().QueryRowContext(ctx, query, id)
+	entity, err := r.mapper.FromRow(row)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, fmt.Errorf("entity not found with id %s: %w", id, err)
+		}
+		return nil, fmt.Errorf("error finding entity with id %s: %w", id, err)
+	}
+
+	return entity, nil
+}
+
+// FindAll retrieves all entities.
+func (r *SQLRepository[T]) FindAll(ctx context.Context) ([]T, error) {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	columns := r.mapper.Columns()
+	// #nosec G202 -- Table name is validated at initialization
+	query := "SELECT " + joinColumns(columns) + " FROM " + r.mapper.Table()
+
+	rows, err := r.db.DB().QueryContext(ctx, query)
+	if err != nil {
+		return nil, fmt.Errorf("error querying all entities: %w", err)
+	}
+	defer rows.Close()
+
+	return r.mapper.FromRows(rows)
+}
+
+// Create inserts a new entity.
+func (r *SQLRepository[T]) Create(ctx context.Context, entity *T) error {
```

Comment:
IMO it would probably be easier to use something like [sqlx](https://github.com/jmoiron/sqlx) which handles most of this mapping and selecting.